fix: Cache and pass person and group properties

haacked opened the pull request:
Fixes #366

## :bulb: Motivation and Context

This fixes a race condition where feature flags return incorrect values when person properties haven't been processed by server-side ingestion yet. Previously, the `/flags` API only sent `distinct_id`, `$anon_distinct_id`, and `$groups` but not person or group properties, causing flags to be evaluated without the latest property values.

This implementation allows person and group properties to be sent directly in feature flag requests, eliminating the timing dependency on async ingestion.

## :green_heart: How did you test it?

- Added new SDK methods `setPersonProperties()` and `setGroupProperties()` with comprehensive documentation
- Updated PostHogExample app with test function demonstrating both person and group properties
- Verified using Proxyman that `person_properties` and `group_properties` are correctly sent in `/flags` API requests
- Tested thread-safe property storage and automatic flag reloading
- Confirmed properties are cleared on `reset()` and properly integrated with `identify()` calls

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

haacked commented:
Linting errors unrelated to this PR. Got a fix for that: https://github.com/PostHog/posthog-ios/pull/374

haritowa commented:
Haha I just though of implementing it :D

marandaneto commented on PostHog/PostHogApi.swift:
```diff
@@ -196,6 +196,8 @@ class PostHogApi {
         distinctId: String,
         anonymousId: String?,
         groups: [String: String],
+        personProperties: [String: Any]? = nil,
```
```suggestion
        personProperties: [String: Any],
```

can probably be non optional since we will attach default person props anyway

marandaneto commented on PostHog/PostHogRemoteConfig.swift:
```diff
@@ -251,9 +257,14 @@ class PostHogRemoteConfig {
                   anonymousId: anonymousId,
-                  groups: groups)
+                  groups: groups,
+                  personProperties: personProperties.isEmpty ? nil : personProperties,
```
I'd remove, less code/less runtime cost

marandaneto commented on PostHog/PostHogRemoteConfig.swift:
```diff
@@ -442,6 +453,55 @@ class PostHogRemoteConfig {
+    func setPersonPropertiesForFlags(_ properties: [String: Any]) {
+        personPropertiesForFlagsLock.withLock {
+            // Merge properties additively, similar to JS SDK behavior
+            for (key, value) in properties {
```
`merging` or `merge` are cleaner, less code.

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +847,146 @@ let maxRetryDelay = 30.0
+    ///
+    /// - Parameter properties: Dictionary of person properties to include in flag evaluation
+    /// - SeeAlso: `setPersonPropertiesForFlags(_:reloadFeatureFlags:)` to control flag reloading behavior
+    @objc public func setPersonPropertiesForFlags(_ properties: [String: Any]) {
```
yes, reason is here https://posthog.com/docs/libraries/js/features#overriding-server-properties

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -468,6 +472,9 @@ let maxRetryDelay = 30.0
             queue.add(event)
 
+            // Automatically set person properties for feature flags during identify event
+            setPersonPropertiesForFlagsIfNeeded(userProperties)
```
we should `sanitizeDictionary(userProperties)` here too since we do that already, we can cache in a var and reuse the result here

marandaneto commented on PostHog/PostHogRemoteConfig.swift:
```diff
@@ -442,6 +453,55 @@ class PostHogRemoteConfig {
+        }
+    }
+
+    func resetGroupPropertiesForFlags(_ groupType: String? = nil) {
```
```suggestion
    func resetGroupPropertiesForFlags(_ groupType: String) {
```

you need a type otherwise there's nothing to do, the caller should avoid the call in this case

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +847,146 @@ let maxRetryDelay = 30.0
+        }
+
+        let sanitizedProperties = sanitizeDictionary(properties) ?? [:]
+        remoteConfig?.setPersonPropertiesForFlags(sanitizedProperties)
```
I'd only call `setPersonPropertiesForFlags` if valid and empty.
Then I'd only `reloadFeatureFlags` inside this condition as well, to not reload for nothing

haacked replied:
Done!

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +847,146 @@ let maxRetryDelay = 30.0
+    /// - Parameter groupType: The group type to clear properties for, or nil to clear all group properties
+    /// - Note: This method does not automatically reload feature flags. Call `reloadFeatureFlags()`
+    ///         after resetting if you want to immediately refresh flags with the cleared properties.
+    @objc public func resetGroupPropertiesForFlags(_ groupType: String? = nil) {
```
```suggestion
    @objc public func resetGroupPropertiesForFlags(_ groupType: String) {
```

marandaneto replied:
@ioannisj do we need a method overload here for better objC uX? otherwise they have to forcefully pass (nil)

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +847,146 @@ let maxRetryDelay = 30.0
+    ///   - properties: Dictionary of person properties to include in flag evaluation
+    ///   - reloadFeatureFlags: Whether to automatically reload feature flags after setting properties
+    @objc(setPersonPropertiesForFlags:reloadFeatureFlags:)
+    public func setPersonPropertiesForFlags(_ properties: [String: Any], reloadFeatureFlags: Bool) {
```
we can make `reloadFeatureFlags = true` as default so you dont have to pass the value, less code.
We should also add an overload for this method for better objC UX only with the `properties` param.

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +847,146 @@ let maxRetryDelay = 30.0
+    /// - Parameters:
+    ///   - properties: Dictionary of person properties to include in flag evaluation
+    ///   - reloadFeatureFlags: Whether to automatically reload feature flags after setting properties
+    @objc(setPersonPropertiesForFlags:reloadFeatureFlags:)
```
```suggestion
    @objc(setPersonPropertiesForFlagsWithProperties:reloadFeatureFlags:)
```

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -468,6 +472,9 @@ let maxRetryDelay = 30.0
             queue.add(event)
 
+            // Automatically set person properties for feature flags during identify event
+            setPersonPropertiesForFlagsIfNeeded(userProperties)
```
the `capture` method also takes the userProps and userSetOnceProps, so we need to call this there too

marandaneto reviewed (approved):
left a few comments, we are on the right direction.

marandaneto reviewed (changes requested):
left a few comments, we are on the right direction.

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -351,6 +351,10 @@ let maxRetryDelay = 30.0
         }
         PostHogSessionManager.shared.resetSession()
 
+        // Clear person and group properties for flags
+        remoteConfig?.resetPersonPropertiesForFlags()
+        remoteConfig?.resetGroupPropertiesForFlags()
```
correct, its not flushing to the disk.
its ok to keep a copy in memory to avoid IO every single time its needed, but right now its not caching at all

haacked replied:
Implemented caching.

marandaneto replied:
@ioannisj https://github.com/PostHog/posthog-ios/pull/373/files#r2224570259
whats about now with caching? do we still need to call these 2 methods?

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -493,6 +503,12 @@ let maxRetryDelay = 30.0
         return false
     }
 
+    private func setPersonPropertiesForFlagsIfNeeded(_ userProperties: [String: Any]?) {
```
should we only allow doing this and group props if `personProfiles` is != `never`?

haacked commented:
@ioannisj @marandaneto thank you so much for your reviews! It's teaching me a lot about mobile development. Would you mind taking another look? I think I addressed all your feedback.

marandaneto commented on PostHog/PostHogRemoteConfig.swift:
```diff
@@ -442,6 +456,70 @@ class PostHogRemoteConfig {
+    }
+
+    private func loadCachedPropertiesForFlags() {
+        personPropertiesForFlagsLock.withLock {
+            if let cachedPersonProperties = storage.getDictionary(forKey: .personPropertiesForFlags) as? [String: Any] {
+                personPropertiesForFlags = cachedPersonProperties
+            }
+        }
+
+        groupPropertiesForFlagsLock.withLock {
+            if let cachedGroupProperties = storage.getDictionary(forKey: .groupPropertiesForFlags) as? [String: [String: Any]] {
+                groupPropertiesForFlags = cachedGroupProperties
+            }
+        }
```
its better the way it is since the `storage` has no locks so we lock the `read` op as well

marandaneto commented on PostHog/PostHogRemoteConfig.swift:
```diff
@@ -442,6 +456,70 @@ class PostHogRemoteConfig {
+    func resetPersonPropertiesForFlags() {
+        personPropertiesForFlagsLock.withLock {
+            personPropertiesForFlags.removeAll()
+            // Clear from disk
+            storage.setDictionary(forKey: .personPropertiesForFlags, contents: personPropertiesForFlags)
```
i think this should be centralized here https://github.com/PostHog/posthog-ios/blob/03803c80ae67695667a8106c98af92caac94e81f/PostHog/PostHogStorage.swift#L352
same for groups

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +847,146 @@ let maxRetryDelay = 30.0
+    ///   - groupType: The group type identifier (e.g., "organization", "team")
+    ///   - properties: Dictionary of properties to set for this group type
+    /// - Note: This method automatically reloads feature flags to apply the new properties.
+    @objc public func setGroupPropertiesForFlags(_ groupType: String, properties: [String: Any]) {
```
still missing

marandaneto replied:
Its needed with method overloads, in this case its not but then the generated method names are different than the others.
Its more about semantics.

marandaneto replied:
If both params would be String, you wouldn't know what comes first

marandaneto commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +862,164 @@ let maxRetryDelay = 30.0
+        guard !sanitizedProperties.isEmpty else { return }
+        remoteConfig?.setGroupPropertiesForFlags(groupType, properties: sanitizedProperties)
+
+        // Automatically reload flags to apply the new properties
+        remoteConfig?.reloadFeatureFlags()
```
should we add a method param similar to `reloadFeatureFlags`? so people can disable if needed

marandaneto reviewed (commented):
left a few final comments, almost there.

haacked commented on PostHog/PostHogSDK.swift:
```diff
@@ -831,6 +862,164 @@ let maxRetryDelay = 30.0
+            return
+        }
+
+        let sanitizedProperties = sanitizeDictionary(properties) ?? [:]
+        guard !sanitizedProperties.isEmpty else { return }
+        remoteConfig?.setPersonPropertiesForFlags(sanitizedProperties)
```
I'll add a check for hasPersonProperties() here.

haacked commented:
Ok, I merged https://github.com/PostHog/posthog-ios/pull/375 into this branch so it all goes in at once.
